### PR TITLE
rewrite spec to be more consistent

### PR DESCRIPTION
**Pull request by sokra**
This PR make as lot of changes to the spec to try to make it more consistent.

One bigger change is how function-imports are handled. This comes with the motivation that people often export arrow functions instead of function declarations. Before this change only function declarations were possible.

In some cases `export type` was renamed to `import type`, as the type of an export in a JS module is not known before evaluation. So decision is now only base on information statically available in WASM.

I also tried to make it more clear that exports are resolved while instantiation and the behavior for reexports is based on that. So reexporting wasm exports in a JS module is not a special case.

I also allowed wasm cycles, but I'm also ok with disallowing this behavior.

### CODE REVIEW

File: proposals/esm-integration/README.md [hunk 1]
```diff
@@ -125,20 +125,21 @@ Some imports are challenging to handle in the current WebAssembly specification.
 
 ### wasm imports <- JS exports
 
-| export type | value (not a WebAssembly.Global)* | global | memory | table | function |
-|-------------|------------------|--------|--------|-------|----------|
-|              | Error                         | Error  | Error  | Error | snapshot |
-
-\*While WebAssembly only has the concept of globals, JS could export either a regular JS value or a `WebAssembly.Global`.
+| import type | global | memory | table | function     |
```
Comment (sendilkumarn) [inline proposals/esm-integration/README.md]: `export`?

File: proposals/esm-integration/README.md [hunk 2]
```diff
@@ -97,18 +97,18 @@ For WebAssembly, instantiation will include initialization for all [external typ
...
+It may be possible to provide live bindings for JS values if an `anyref` type is added to WebAssembly. At this point the restriction prohibiting non-function imports could also be lifted.
 
 The reason we want to throw an error in this case is two fold:
 
-- **Reduce confusion for developers.** Since non-function imports work in JS->JS dependency chains, it is unintuitive that they would be undefined in JS->wasm dependency chains. Throwing an error makes clear what's happening.
-- **Allow for future support.** It may be possible to support non-function imports to wasm modules if an `any` type is added. We close off this avenue if we allow users to import values that can be undefined, though, as code may start depending on that behavior.
+- **Reduce confusion for developers.** WebAssembly snapshots imports while instantiation, but most exports from JS modules are still `undefined` or in TDZ at this time. Since non-function imports work in JS->JS dependency chains, it is unintuitive that they would be undefined in JS->wasm dependency chains. Throwing an error makes clear what's happening.
```
Comment (littledan) [inline proposals/esm-integration/README.md]: This snapshotting time is the proposal at master, and this paragraph ("shapshots imports while instantiation") is characteristic of it, but I like the PR at #13, which changes the time to be a bit later (during execution) and allows more cases to be well-handled.

File: proposals/esm-integration/README.md (hunk 1 above)
Comment (littledan) [reply proposals/esm-integration/README.md]: import type seems more appropriate here--the path is chosen based on how the WebAssembly module declares its imports. I really appreciate this fix.

File: proposals/esm-integration/README.md [hunk 3]
```diff
@@ -125,20 +125,21 @@ Some imports are challenging to handle in the current WebAssembly specification.
...
-1. JS module is instantiated. Functions are initialized. Non-function imports are set to undefined.
-1. wasm module is instantiated. A snapshot is taken of all imports. Functions are copied. All non-function imports are undefined and thus throw an error.
-
-Errors will happen automatically because the imports would be undefined, and WebAssembly currently does not support importing undefined values.
+1. Construction
+  1. wasm module is parsed
+  1. JS module is parsed
+1. Instantiation
+  1. JS module is instantiation and bindings are created to exports.
+  1. wasm module is instantiated. A trampoline functions is created for function imports, which points to the exported binding. All non-function imports throw an error.
```
Comment (littledan) [inline proposals/esm-integration/README.md]: Let's consider these trampolines just when needed to handle the circular case, in a follow-on proposal, see #17

File: proposals/esm-integration/README.md [hunk 4]
```diff
@@ -159,7 +160,7 @@ function getCount() {
 export {getCount};
 ```
 
-##### Value imports
+##### Global imports
```
Comment (littledan) [inline proposals/esm-integration/README.md]: Eh, from the perspective of a JS programmer, "value" makes sense.

**Comment by littledan** [comment]
I think we'll go with the semantics checked in right now, rather than the semantics in this PR, due to the inability to handle importing Memory, Tables, and in the future, nominal types.

**Pull request closed without merging**